

# OUTSIDER

**OUTSIDER** are TE no assembled on genome

## MAPPING

Information `samtools stats`

File : **OUTSIDER/MAPPING/stats.txt**

```{bash MAPPING_STATS_OUTSIDER, eval=FALSE, include=FALSE, results=TRUE}
    cat "../../OUTSIDER/MAPPING/stats.txt"
```

```{r show_STATS, eval=TRUE, include=TRUE, results=TRUE}
    print(gsub("\t" , "    ", readLines("../../OUTSIDER/MAPPING/stats.txt")))
```

## SV ANALYSIS

Informations of **SV.vcf** file.

```{bash COUNT_SV_OUTSIDER, eval=TRUE, include=TRUE, results=TRUE}

array_type=( "INS" "DEL" "DUP" ); 
echo -e "x\ty" > .tmp_graph_sv.csv; 
for i in ${array_type[@]}; do 
    NB=`grep -w $i ../../OUTSIDER/VARIANT_CALLING/SV.vcf | cut -f 2 | sort | uniq | wc -l || echo 0`;
    echo -e "$i\t$NB";
    sed -i "s/%%VALUE_${i}%%/${NB}/g" lib/template_HISTO_TYPE_SV.html
done >> .tmp_graph_sv.csv;

NB_INS_TrEMOLO=`cat ../../OUTSIDER/TrEMOLO_SV_TE/INS/SV_INS_CLUST.bed | wc -l || echo 0`
sed -i "s/%%VALUE_INS_TrEMOLO%%/${NB_INS_TrEMOLO}/g" lib/template_HISTO_TYPE_SV.html


NB_SOFT=`cat ../../OUTSIDER/TrEMOLO_SV_TE/SOFT/SV_SOFT.vcf | wc -l || echo 0`
sed -i "s/%%VALUE_SOFT%%/${NB_SOFT}/g" lib/template_HISTO_TYPE_SV.html


NB_HARD=`(test -s ../../OUTSIDER/TrEMOLO_SV_TE/HARD/HARD.fasta && grep ">" ../../OUTSIDER/TrEMOLO_SV_TE/HARD/HARD.fasta  | cut -d ":" -f 5 | sort -u | wc -l) || echo 0`
sed -i "s/%%VALUE_HARD%%/${NB_HARD}/g" lib/template_HISTO_TYPE_SV.html

```

<iframe src="lib/template_HISTO_TYPE_SV.html" style="width: 100%; height: 400px; border: none; overflow: hidden;"></iframe>

## NUMBER OF TE

Total number of **TE : `r data=read.csv("../../OUTSIDER/TE_DETECTION/MERGE_TE/MERGE_TE_ALL.bed", sep="\t", header=FALSE); nrow(data);`**

Number of family found : **`r data=read.csv("../../OUTSIDER/TE_DETECTION/MERGE_TE/MERGE_TE_ALL_COUNT.csv", sep="\t"); nrow(data);`**

```{r NB_OUTSIDER, eval=TRUE, include=TRUE, results=TRUE, out.width="90%"}

    data = read.csv("../../OUTSIDER/TE_DETECTION/MERGE_TE/MERGE_TE_ALL_COUNT.csv", sep="\t")
    #kbl(data)
    data = tail(data, 50)

    uniqy = unique(data$y)
    uniqx = unique(data$x)

    for (y in uniqy){
        datay = data[data$y %in% y, ]
        for(x in uniqx){
            datax = datay[datay$x %in% x, ]
            if(nrow(datax) != 1){
                new_line_data <- data.frame(x, y, 0)
                names(new_line_data) <- c("x", "y", "z")

                data <- rbind(data, new_line_data)
            }
        }
    }


    max=0
    namx=""
    for(x in uniqx){
        datax = data[data$x %in% x, ]
        if(max(datax$z)>max){
            max  = max(datax$z)
            namx = x
        }
    }

    datax = data[data$x %in% namx, ]
    datax = datax[order(datax$z), ]

    data$y <- factor(data$y, levels = datax$y)

    nb_diff_TE = length( unique(data$y) )

    xlabel = ''
    graph <- ggplot(data, aes(x, y, fill=z, width=0.95, height=0.95)) + 
        geom_tile() + 
        ggtitle("") +
        theme(axis.text.x = element_blank(), axis.text.y = element_text(face="plain", color="#222222", size=5, angle=0)) +
        coord_fixed() +
        scale_fill_gradientn(colours=brewer.pal(n= 9, name="Reds")) +
        #scale_fill_viridis(option="heat") +
        theme(panel.background = element_rect(fill = "white", colour = "grey", linewidth=0)) +
        geom_segment( aes(x = 5, y = 1, xend = 5, yend = nb_diff_TE ), colour = "black", alpha=0, linewidth=0.9 , inherit.aes = FALSE , linetype="twodash") +
        labs(y="TE", x=xlabel, fill="NUMBER OF TE") +
        geom_text(aes(label = z), nudge_x=1.2, color = "#222222", size = 2) 

    graph

```

<iframe src="lib/template_COUNT_TE_OUTSIDER.html" style="width: 100%; height: 1200px; border: none; overflow: hidden;"></iframe>

<iframe src="lib/template_COUNT_TE_SENS_ANTISENS_OUTSIDER.html" style="width: 100%; height: 800px; border: none; overflow: hidden;"></iframe>

## FREQUENCY

<iframe src="lib/template_FREQ_TE_OUTSIDER.html" style="width: 100%; height: 1800px; border: none; overflow: hidden;"></iframe>

Hover your mouse over the bars or navigate with the slider if the number of different TE families is **greater than 20**

<iframe src="lib/template_FREQ_CHROM_OUTSIDER.html" style="width: 100%; height: 100vh; border: none; overflow: hidden;"></iframe>


## TSD

<!-- calcul number of TE TSD -->


```{bash COUNT_TSD_OUTSIDER, eval=TRUE, include=TRUE, results=TRUE}

grep "OK" ../../VALUES_TSD_GROUP_OUTSIDER.csv | awk 'BEGIN{somme=0}{somme+=$2}END{print somme}' > .tmp_TE_TSD_OK.txt || echo "ERROR TSD NOT FOUND"

grep "KO" ../../VALUES_TSD_GROUP_OUTSIDER.csv | awk 'BEGIN{somme=0}{somme+=$2}END{print somme}' > .tmp_TE_TSD_KO.txt || echo

awk 'BEGIN{somme=0} NR>1 {somme+=$2}END{print somme}' ../../VALUES_TSD_GROUP_OUTSIDER.csv > .tmp_TE_TSD_TOTAL.txt || echo

```

TSD [OK] : these are the TE with TSD **`r readLines(".tmp_TE_TSD_OK.txt");`/`r readLines(".tmp_TE_TSD_TOTAL.txt");`**

TSD [KO] : these are the TE without TSD **`r readLines(".tmp_TE_TSD_KO.txt");`/`r readLines(".tmp_TE_TSD_TOTAL.txt");`**


```{r TSD, eval=FALSE, include=TRUE, results=FALSE, out.width="90%", echo=FALSE}

data = read.csv("../../VALUES_TSD_GROUP_OUTSIDER.csv", sep="\t")
data = tail(data, 50)

data$x <- factor(data$x, levels = order_TE)

p = ggplot(data, aes(fill=condition, y=y, x=x)) + 
        geom_bar(position="stack", stat="identity") +
        xlab("TE") +
        ylab("NUMBER") +
        labs(fill = "TYPE TSD") +
        #scale_fill_viridis(discrete = T) +
        scale_fill_manual(values=c("#6EB9FF", "#0064D4")) +
        theme(
            axis.text.x = element_text(face="plain", color="#222222", size=10, angle=90, vjust=0.5),
            panel.background = element_rect(
                    fill="white", 
                    colour="white", 
                    linetype="solid", 
                    color="white"
                )
            )

p
#ggsave(p, file=output)

```

<!-- <iframe src="lib/template_TSD_OUTSIDER_histo.html" style="width: 100%; height: 800px; border: none; overflow: hidden;"></iframe> -->
<iframe src="lib/template_TSD_OUTSIDER_CHART.html" style="width: 100%; height: 800px; border: none; overflow: hidden;"></iframe>

Hover your mouse over the bars or navigate with the slider if the number of different TE families is **greater than 20**

<!-- <iframe src="lib/template_TSD_ALL_histo.html" style="width: 100%; height: 700px; border: none; overflow: hidden;"></iframe> -->


<!-- remove inutiles files -->
```{bash rm_OUTSIDER, eval=TRUE, include=TRUE, results=TRUE}

rm -f FREQ_TE.csv
rm -f .tmp_*v;

```



