# Pipeline minority variants

*******
Content 
 1. [Configuration of the parameter file.](#config_file)
 2. [Start the pipeline.](#start_pipeline)

*******

## Require 

  * [Snakemake](https://snakemake-wrappers.readthedocs.io/en/stable/) 5.5.2
  * [Minimap2](https://github.com/lh3/minimap2) 2.16+
  * [Samtools](http://www.htslib.org/) 1.10+
  * [Sniffles](https://github.com/fritzsedlazeck/Sniffles) 1.0.10+
  * [Biopython](https://biopython.org/)
  * [Pandas](https://pandas.pydata.org/)
  * [Numpy](https://numpy.org/)

<div id='config_file'/> 

## Configuration of the parameter file 

You will first have to enter your parameters in a **.json** file (see example config.json file). The necessary parameters are:

```
{
    "work_directory": "/path/to/directory_work",
    "genome": "/path/to/genome_file.fasta",
    "read": "/path/to/reads_file.fastq",
    "name_out": "prefix_of_name_files_out",
    "fasta_TE": "/path/to/database_TE.fasta"
}
```

ğŸ‘‰ ***You can pass absolute or relative paths.*** ğŸ‘

* **work_directory** : directory that will contain the output files (if the directory does not exist it will be created).
* **genome** : Genome assembly
* **read** : File containing the reads of genome assembly.
* **name_out** : Choice of the prefix or suffix of the majority of the output files.
* **fasta_TE** : File **.fasta** containing the sequence of transposable elements.

<div id='start_pipeline'/> 

## Start the pipeline

```
snakemake --snakefile /path/to/vrare.snk --configfile /path/to/your_config.json
```

ğŸ‘‰ ***You can pass absolute or relative paths.*** ğŸ‘


### Summarize output files :open_file_folder:

Example of output file obtained after using the pipeline, in work directory.

```
â”œâ”€â”€ ET_FIND_FA
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_17.6.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_297.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_412.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_blood.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_Burdock.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_copia.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_diver.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_flea.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_gtwin.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_HMS-Beagle.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_Max-element.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_mdg3.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_opus.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_Quasimodo.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_roo.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_rover.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_springer.fasta
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_find_Transpac.fasta
â”‚Â Â  â””â”€â”€ G73vsG73LR_find_ZAM.fasta
â”œâ”€â”€ FASTA_FIND
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10077:10091462-10098788.fasta
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10077:10091462-10098788.fasta.fai
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10077:10091462-10098788.fasta.nhr
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10077:10091462-10098788.fasta.nin
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10077:10091462-10098788.fasta.nsq
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:1024:897326-904652.fasta
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:1024:897326-904652.fasta.fai
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:1024:897326-904652.fasta.nhr
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:1024:897326-904652.fasta.nin
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:1024:897326-904652.fasta.nsq
â”‚Â Â  â”œâ”€â”€ ...
â”œâ”€â”€ G73vsG73LR.bam
â”œâ”€â”€ G73vsG73LR_cnTE_ALL_ET.csv
â”œâ”€â”€ G73vsG73LR_cnTE_ALL_ET_hit_map.csv
â”œâ”€â”€ G73vsG73LR_cnTE.bln
â”œâ”€â”€ G73vsG73LR_MD.sorted.bam
â”œâ”€â”€ G73vsG73LR_sniffle.fasta
â”œâ”€â”€ G73vsG73LR_sniffle.fasta.fai
â”œâ”€â”€ G73vsG73LR.sorted.bam
â”œâ”€â”€ G73vsG73LR_total_find.fasta
â”œâ”€â”€ G73vsG73LR.vcf
â”œâ”€â”€ Gene_G73onG0corrected_ragoo_full.bln
â”œâ”€â”€ HISTO
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_cnTE_ALL_ET_17.6_histo.pdf
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_cnTE_ALL_ET_297_histo.pdf
â”‚Â Â  â”œâ”€â”€ G73vsG73LR_cnTE_ALL_ET_412_histo.pdf
â”‚Â Â  â”œâ”€â”€ ...
â”œâ”€â”€ PICLUSTER
â”‚Â Â  â”œâ”€â”€ G73onG0corrected_ragoo_IN_CLUSTER_COV_hit_map.csv
â”‚Â Â  â”œâ”€â”€ Gene_G73onG0corrected_ragoo_combine.csv
â”‚Â Â  â””â”€â”€ PiCluster_Gene_G73onG0corrected_ragoo.csv
â”œâ”€â”€ READ_FASTQ_G73vsG73LR
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10077:10091462-10098788.fastq
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:1024:897326-904652.fastq
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10273:10309986-10317318.fastq
â”‚Â Â  â”œâ”€â”€ ...
â”œâ”€â”€ REGION_RD_G73vsG73LR
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10077:10091462-10098788.txt
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:1024:897326-904652.txt
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:10273:10309986-10317318.txt
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:11769:12063423-12070758.txt
â”‚Â Â  â”œâ”€â”€ reads_2L_RaGOO_RaGOO:11994:12341312-12349689.txt
â”‚Â Â  â”œâ”€â”€ ...
â””â”€â”€ TSD
    â”œâ”€â”€ total_results_tsd_17.6.txt
    â”œâ”€â”€ total_results_tsd_297.txt
    â”œâ”€â”€ total_results_tsd_412.txt
    â”œâ”€â”€ ...
```

##### Diagram of the different stages (rules)

<img src="dag.svg">

Enjoy :+1:


